staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -319,12 +349,22 @@ func (w *watchCache) waitUntilFreshAndBlock(resourceVersion uint64, trace *utilt
 }
 
 // WaitUntilFreshAndList returns list of pointers to <storeElement> objects.
-func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, trace *utiltrace.Trace) ([]interface{}, uint64, error) {
+func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, matchValues []storage.MatchValue, trace *utiltrace.Trace) ([]interface{}, uint64, error) {
 	err := w.waitUntilFreshAndBlock(resourceVersion, trace)
 	defer w.RUnlock()
 	if err != nil {
 		return nil, 0, err
 	}
+
+	// This isn't the place where we do "final filtering" - only some "prefiltering" is happening here.
+	// So the only requirement here is to NOT miss anything that should be returned. We can return as
+	// many non-matching items as we want - they will be filtered out later. The fact that we return less
+	// things is only further performance improvement.
```

Comment:
`// TODO: if multiple indexes match, return the one with the fewest items, so as to do as much filtering as possible.`?

Reply:
done